feat: uniswap plugin and example

dannythedawger opened the pull request:
## Description

- This PR aims to introduce the skeleton of a Uniswap plugin
- Lints the repository, hence the diff on a few other plugins
- Adds an example for use of the plugin using *vercel ai*

dannythedawger commented on typescript/packages/plugins/uniswap/src/tools.ts:
```diff
@@ -0,0 +1,49 @@
+                return getSwapTransaction(parameters, apiKey, baseUrl);
+            },
+        },
+        {
+            name: "send_transaction",
+            description: "This {{tool}} signs and sends a transaction.",
+            parameters: SendTransactionBodySchema,
+            method: async (walletClient: EVMWalletClient, parameters: z.infer<typeof SendTransactionBodySchema>) => {
+                return sendTransaction(parameters, walletClient);
+            },
+        },
+    ];
```
had to create this one as I didn't see a wallet plugin that can send a primitive tx

0xaguspunk replied:
A generic sendTx function is quite complex for an LLM to figure out what to do with it.

The goal with these plugins is to have specific functions (e.g swapTokens, createLiquidityPool, collectFees, etc.)

Then these functions can use sendTransaction internally.

I would remove this one. There is another dev adding those specific calls in a different PR so maybe they could contribute to this base with those functions :)

0xaguspunk replied:
That is the case for AI SDK but other AI agent frameworks are not there yet. In any case let's go with this approach! :)

Could you move the sendTransaction method [here](https://github.com/goat-sdk/goat/tree/main/typescript/packages/core/src/plugins) instead? Make it a core plugin instead.

0xaguspunk commented on typescript/packages/plugins/uniswap/src/index.ts:
```diff
@@ -0,0 +1,24 @@
+    return {
+        name: "Uniswap",
+        supportsChain: (chain: Chain) => chain.type === "evm" && SUPPORTED_CHAINS.some((c) => c.id === chain.id),
+        supportsSmartWallets: () => false,
```
Based on the code it does seem like this plugin would support smart wallets! :)

dannythedawger replied:
Yep, good callout!

0xaguspunk commented on typescript/packages/plugins/uniswap/src/tools.ts:
```diff
@@ -0,0 +1,49 @@
+export function getTools({ apiKey, baseUrl }: UniswapToolsOptions): DeferredTool<EVMWalletClient>[] {
+    return [
+        {
+            name: "check_approval",
```
There is already the approval action on the erc20 plugin. I would leverage that one and remove this one from here. So I can install both erc20 and uniswap plugins and the agent can use both to achieve their goal

dannythedawger replied:
this is not a generic approval call, it's part of the Uniswap Trading API suite - it does approvals for all sort of functionality including intents and bridging, with and without Permit2. The plugin is not usable without this.

0xaguspunk commented on typescript/packages/plugins/uniswap/README.md:
```diff
@@ -0,0 +1,34 @@
+
+const plugin = uniswap({
+    baseUrl: process.env.UNISWAP_BASE_URL as string,
+    apiKey: process.env.UNISWAP_API_KEY as string,
```
Where does one get a Uniswap API key? It'd be great to link to that on the readme

dannythedawger replied:
Ah I included the link to the landing page and form in the *vercel ai* example I created - shall I add it here too?

0xaguspunk commented on typescript/examples/vercel-ai/uniswap/.env.template:
```diff
@@ -0,0 +1,3 @@
+# Uniswap (get an API key by running the `pnpm create-api-key` command)
```
Is this script pushed?

dannythedawger replied:
Nope, my but - I copied it over from a template I used

0xaguspunk reviewed (changes requested):
Neat! Would love to get this merged! Made a few comments on:
1. Functions we might not need
2. Support for smart wallets
3. Getting an API key to use this

Otherwise, lgtm!!

dannythedawger commented:
@0xaguspunk thansk for great feedback! I addressed most of the comments and left 1 open question to discuss further. We should also introduce lint checks before merging as the bulk of my PR is linting other pkg's code!

0xaguspunk commented:
Neat 🔥 Will take a look at it in the morning! Agree with the lint, so annoying, haha will look into that too!

0xaguspunk reviewed (approved):
This is great 🔥

Last (sorry if I'm asking too much!): maybe rebase from main and run `pnpm change:add` to make sure we release this! :)

dannythedawger commented:
@0xaguspunk thanks for the approval, addressed your last comment too! For some reason, the tests won't run

0xaguspunk reviewed (approved):
Two small things:
1. Could we call it `eth-send-transaction.ts` to make it consistent?
2. Does it need to be exported from the `index.ts` file in the root of `core`?

Let's merge it after this!

dannythedawger commented:
> Two small things:
> 
> 1. Could we call it `eth-send-transaction.ts` to make it consistent?
> 2. Does it need to be exported from the `index.ts` file in the root of `core`?
> 
> Let's merge it after this!

Should be ready to merge - again, the build and test command are not running for some reason

0xaguspunk commented:
hmm weird. Will merge and take a look at it
